core/src/main/scala/filodb.core/downsample/DownsampleCardinalityManager.scala

```diff
@@ -0,0 +1,170 @@
+package filodb.core.downsample
+
+import com.typesafe.config.Config
+import com.typesafe.scalalogging.StrictLogging
+
+import filodb.core.DatasetRef
+import filodb.core.memstore.PartKeyLuceneIndex
+import filodb.core.memstore.ratelimit.{CardinalityTracker, QuotaSource, RocksDbCardinalityStore}
+import filodb.core.metadata.PartitionSchema
+import filodb.core.store.StoreConfig
+import filodb.memory.format.UnsafeUtils
+
+/**
+ * Manages the calculation and storage of cardinality count in a shard of downsample cluster
+ */
+class DownsampleCardinalityManager(rawDatasetRef: DatasetRef,
+                                   shardNum: Int,
+                                   shardKeyLen: Int,
+                                   partKeyIndex: PartKeyLuceneIndex,
+                                   partSchema: PartitionSchema,
+                                   filodbConfig: Config,
+                                   downsampleStoreConfig: StoreConfig,
+                                   quotaSource: QuotaSource) extends StrictLogging {
+
+  var cardTracker: CardinalityTracker = initCardTracker()
+
+  // number of shards per node - used to stagger/distribute the cardinality count calculation of different shards
+  // to different times of day
+  private val numShardsPerNode: Int = getNumShardsPerNodeFromConfig(rawDatasetRef.dataset, filodbConfig)
+
+  /**
+   * Helper method to get the shards per node in the downsample filodb cluster. We look for the two specific
+   * config that is `min-num-nodes` ( the minimum number of nodes/pods which has to be in a cluster ) and
+   * `num-shards` (the total number of shards). num-shards/min-num-nodes gives us the required value.
+   *
+   * NOTE: This configs can be part of `dataset-configs` or `inline-dataset-configs`. Also making sure, we look
+   * for the dataset, which was used to create this instance of DownsampleTimeSeriesShard
```

Comment:
What is the conceptual difference between these two configs? What does "inline" imply?

Reply:
So currently,  filodb-default configs have two sections where a dataset's config could be present. They are `dataset-configs` and `inline-dataset-configs`. That's why we are checking for both this sections of configs